keps/sig-api-machinery/1904-efficient-watch-resumption/README.md

```diff
@@ -0,0 +1,502 @@
+and kube-apiserver replicas, kube-apiserver contains an in-memory cache
+called `watch cache` from which all watch requests as well as non-quorum
+reads (in opt-in fashion) are served. Watch cache is being propagated
+directly from etcd using typical "list + watch" pattern and stores both
+the current state as well as som recent 'transaction log'. There is a
+separate watch cache for each resource type version.
+
+Digging into more details, the way how watch cache is propagated is:
+- on start, quorum list is send to etcd
+- given the RV of the list represents the current "global etcd version",
+  it doesn't have to reflect any change of objects of our type
+- the watch cache is set to be synced to list resource version (which
+  as mentioned above isn't necessary reflecting any change of objects
+  of that type)
+- from the on, watch to etcd is established and watch cache is being
+  updated by incoming watch event stream; note that only objects of
+  a given resource type are being watched and only those can update
+  the resource version to which watch cache is synced from then on
+- if watch cannot be reestablished, all watchers connected to it are
+  disconnected, a new quorum list is send and watchcache is reset to
+  that point in time
+
+It's important to note that while only objects of a given resource type
+are watched (thus resource version to which watch cache is synced can
+only be updated to a value that some object of that type had), the quorum
+list, even though it is only requesting objects of that resource type,
+returns "globally current version for the whole etcd".
+
+While the inability to reestablish watch doesn't happen often in practice,
+the kube-apiserver restarts obviously does (e.g. on upgrades). And this
+is resulting in two main problems:
+
+1. In case of rolling upgrade of kube-apiserver, when no object of a given
+   resource type is changing (but objects of other types do), all watchers
+   will eventually be forced to relist, causing significant performance
+   and scalability issues for larger clusters. See the example below for
+   clarification.
+
+1. If no object of a given resource type is changing after kube-apiserver
+  startup, watch cache may stuck being synced to resource version not being
+	version of any object of a given type for extended period of time. See
+  https://github.com/kubernetes/kubernetes/issues/91073 for more details.
+
+To illustrate the problem better, imagine the following example:
+- there is a single kube-apiserver that we are going to restart
+- there are only two resource types: X and Y (for simplicity)
+- there is one object of type X: x with rv=100
+- there is one object of type Y: y with rv=101
+- there is a watch W for objects of type X already synced to rv=100
+- watchcache for both X and Y is up-to-date
+
+- T1: kube-apiserver being restarted
+- T2: kube-apiserver is initialized with watchcache for X synced to
```

Comment:
it initialized this based on a list, correct?  if so, please indicate that inline.